sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/GoneAndRetryWithRetryPolicy.java

```diff
@@ -15,190 +14,347 @@
-                                exception, this.lastRetryWithException);
-                        exceptionToThrow = this.lastRetryWithException;
-                    } else {
-                        logger.warn(
-                                "Received partition key range gone exception after backoff/retry. Will fail the request. {}",
-                                exception.toString());
-                        exceptionToThrow = BridgeInternal.createServiceUnavailableException(exception);
-                    }
-                } else if (exception instanceof InvalidPartitionException) {
-                    if (this.lastRetryWithException != null) {
-                        logger.warn(
-                                "Received InvalidPartitionException after backoff/retry including at least one RetryWithException. "
-                                        + "Will fail the request with RetryWithException. InvalidPartitionException: {}. RetryWithException: {}",
-                                exception, this.lastRetryWithException);
-                    } else {
-                        logger.warn(
-                                "Received invalid collection partition exception after backoff/retry. Will fail the request. {}",
-                                exception.toString());
-                        exceptionToThrow = BridgeInternal.createServiceUnavailableException(exception);
-                    }
-                } else {
-                    logger.warn("Received retrywith exception after backoff/retry. Will fail the request. {}",
-                            exception.toString());
+
+        @Override
+        public Mono<ShouldRetryResult> shouldRetry(Exception exception) {
+            CosmosException exceptionToThrow;
+            Duration backoffTime = Duration.ofSeconds(0);
+            Duration timeout;
+            boolean forceRefreshAddressCache;
+            if (!isRetryableException(exception)) {
+
+                logger.debug("Operation will NOT be retried. Current attempt {}, Exception: ", this.attemptCount,
+                    exception);
+                return Mono.just(ShouldRetryResult.noRetry());
+            } else if (exception instanceof GoneException &&
+                !request.isReadOnly() &&
+                BridgeInternal.hasSendingRequestStarted((CosmosException)exception)) {
+
+                logger.warn(
+                    "Operation will NOT be retried. Write operations can not be retried safely when sending the request " +
+                        "to the service because they aren't idempotent. Current attempt {}, Exception: ",
+                    this.attemptCount,
+                    exception);
+
+                return Mono.just(ShouldRetryResult.noRetry(
+                    Quadruple.with(true, true, Duration.ofMillis(0), this.attemptCount)));
+            }
+
+            long remainingSeconds = this.waitTimeInSeconds - this.getElapsedTimeSupplier.get().toSeconds();
```

Comment:
Duration.toSeconds() is a java9 + api. doesn't exist on java8
https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/time/Duration.html#toSeconds()

won't work on java8

```suggestion
            long remainingSeconds = this.waitTimeInSeconds - (this.getElapsedTimeSupplier.get().toMillis() / 1000);
```

Reply:
Good catch :-)

Reply:
Fixed